s390x: update some regalloc metadata to remove use of `reg_mod`.

cfallin opened the pull request:
This is a step toward ultimately removing modify-operands, which along
with removal of pinned vregs, lets us move to a completely
constraint-based and fully-SSA regalloc input and get some nice
advantages eventually.

There are still a few uses of `mod` operands and pinned vregs remaining,
especially around the "regpair" abstraction. Those proved to be a bit
trickier to update though, so will have to be done separately.

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

cfallin commented:
cc @uweigand to review?

I also spent the past day trying to go further in [this branch](https://github.com/cfallin/wasmtime/tree/s390x-ra2-semantics-2), and managed to clean up a lot more, but there are some panics from undefined regs there so I don't have it quite right. @uweigand if you have time to poke at this more I'd very much appreciate it (but it's not urgent at all!). My long-term goal is to push the regalloc input toward fully-SSA (no mods, no multiple defs) and fully constraint-based (no pinned vregs) input, which allows for more flexibility in managing copies and spills, and makes for a more efficient solver generally.

uweigand commented on cranelift/codegen/src/isa/s390x/inst.isle:
```diff
@@ -2702,7 +2713,7 @@
 (rule (push_alu_uimm32shifted ib op (real_reg dst) r imm)
       (if (same_reg dst r))
-      (let ((_ Unit (inst_builder_push ib (MInst.AluRUImm32Shifted op dst imm))))
+      (let ((_ Unit (inst_builder_push ib (MInst.AluRUImm32Shifted op dst dst imm))))
```
In fact, the whole `same_reg` logic can probably go away then, as the assertion that the two register are actually tied is now made during `emit` anyway.

cfallin replied:
Done!

cfallin commented on cranelift/filetests/filetests/isa/s390x/arithmetic.clif:
```diff
@@ -21,7 +21,7 @@ block0(v0: i64, v1: i64):
 
 ; block0:
-;   agr %r2, %r3
+;   agr %r2, %r2, %r3
```
This is a great idea actually; thanks! It nicely solves the issue of showing all info when viewing pre-regalloc VCode, but showing machine code as it really is once regalloc occurs. I've made this change, and I plan to adopt it to the mod operands in the other backends as well.

cfallin commented on cranelift/codegen/src/isa/s390x/inst/emit.rs:
```diff
@@ -1426,7 +1426,12 @@ impl MachInstEmit for Inst {
+                    let inst = Inst::AluRR {
+                        alu_op,
+                        rd,
+                        ri: rd.to_reg(),
```
Updated to use `rn`'s value, thanks. I actually lean slightly toward keeping these fields named `ri`, to make it clear that they are artificial instruction fields, the "input" side of the dest reg, rather than a real `rn` field (this also made it easier to grep for things when updating tests just now!). But I'm happy to alter the field name as well if you feel strongly about this.

cfallin commented:
Thanks @uweigand! I've updated based on feedback (and, importantly, reverted the 2-to-3-arg change in assembly printing). Thanks for looking further at the followup patch as well; I will pick that up and try to finish it next week, most likely.

uweigand commented on cranelift/filetests/filetests/isa/s390x/bitops.clif:
```diff
@@ -36,34 +36,36 @@ block0(v0: i64):
+;   lgr %r3, %r2
+;   llihf %r2, 2863311530
+;   iilf %r2, 2863311530
+;   lgr %r5, %r3
```
These two `lgr` look new - this is why the new code is two instructions longer than the old code.  Not sure if this is something that could still be improved in regalloc, or if this is just one of those random changes ...   In any case, not a big deal, I just wanted to point it out.

cfallin replied:
I briefly looked but it was nothing really obvious. It's possible that my changes in bytecodealliance/regalloc2#74 might help a bit, but I'm not sure; let's see if it reverts back once I update tests there with this merged :-)

uweigand reviewed (approved):
LGTM now.   Note the inline comment about one minor regalloc regression, but that doesn't block this PR.

cfallin commented:
Ah, I think this needs an r+ from someone with write access to the repo -- anyone want to give a rubberstamp on top of Ulrich's review above?
